### ISSUE TITLE
urllib.request resolves the host before checking it against the system's proxy bypass list [Security: LOW, minor info leak]

### ISSUE THREAD

**Issue by weiiwang01** [python/cpython#115197]
# Bug report

### Bug description:

When system proxy bypass list is set, the urllib.request library on macOS and Windows resolves the hostname to an IP address and the IP address to a hostname (on Windows) before checking it against the system proxy bypass list (see [here](https://github.com/python/cpython/blob/5914a211ef5542edd1f792c2684e373a42647b04/Lib/urllib/request.py#L2589-L2594) and [here](https://github.com/python/cpython/blob/5914a211ef5542edd1f792c2684e373a42647b04/Lib/urllib/request.py#L2731-L2743)).

This causes DNS leak and HTTP requests to hang while waiting for DNS timeout in some air-gaped environments. This behavior also differs from other system applications (tested on macOS Sonoma with Safari and Windows Server 2022 with the Edge browser).

<details>
<summary>Test process on macOS and Windows:</summary>

Creating an A record from `<my-test-domain>.net` to `<my-test-ip>`.

macOS with Safari:

In the system network setting:
- "Web proxy (HTTP)" is set to 172.16.0.1:8000
- "Secure web proxy (HTTPS)" is set to 172.16.0.1:8000
- "Bypass proxy settings" is set to `<my-test-ip>`

In Safari:
- visiting `http://<my-test-ip>`: does not use the proxy
- visiting `http://<my-test-domain>.net`: uses the proxy

Windows Server 2022 with Edge browser:

in system network setting:
- "HTTP proxy" is set to 172.16.0.1:8000
- "Do not use proxy server" is set to `<my-test-ip>`

In Edge browser:
- visiting `http://<my-test-ip>`: does not use the proxy
- visiting `http://<my-test-domain>.net`: uses the proxy

urllib.request on Windows also resolves the IP address back to FQDN before check, here's a test for that:

Windows Server 2022 with Edge browser:

Update the Host file so the IP address can be resolved back to FQDN (`socket.getfqdn("<my-test-ip>") == "<my-test-domain>.net"`).

In system network setting:
- "HTTP proxy" is set to 172.16.0.1:8000
- "Do not use proxy server" is set to <my-test-domain>.net

In Edge browser:
- visiting `http://<my-test-ip>`: uses the proxy
- visiting `http://<my-test-domain>.net`: does not use the proxy

</details>

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

macOS, Windows

### Linked PRs
* gh-115210
* gh-116066
* gh-116067
* gh-116068
* gh-116069
* gh-116070

**Comment by weiiwang01** [comment]
I have created a fix for this problem internally. If it's okay, I can submit that as a pull request.

**Comment by zooba** [comment]
Yes, please add a pull request. If it's a relatively simple fix, then this will be less controversial.

For what it's worth, we don't usually consider web browsers to be the canonical behaviour for our APIs. If another low-level library similar to urllib is doing it, that is helpful guidance (e.g. urllib3).

**Comment by weiiwang01** [comment]
> Yes, please add a pull request. If it's a relatively simple fix, then this will be less controversial.
> 
> For what it's worth, we don't usually consider web browsers to be the canonical behaviour for our APIs. If another low-level library similar to urllib is doing it, that is helpful guidance (e.g. urllib3).

@zooba Yes, the solution is simply not to resolve the host before checking it against the system proxy bypass list. I don't recall any HTTP tools, for example, curl, has this behavior. I did some research and found that the `requests` library has already created [a patch inside `requests`](https://github.com/psf/requests/commit/1c38e1f5f64c5da8a39a890d9205dd49297a9914) to stop resolving for Windows, but it is still unfixed for macOS. Fixing this behavior in Python could definitely help.

**Comment by weiiwang01** [comment]
> most

@ronaldoussoren Thank you for the investigation. This aligns more or less with what I thought. Even Python itself does not resolve the hostname when handling the `no_proxy` environment variable, just like `wget`, `go`, and `curl`. https://github.com/python/cpython/blob/31633f4473966b3bcd470440bab7f348711be48f/Lib/urllib/request.py#L2516-L2548

The affected code here is slightly different, as Python handles the proxy bypass list sourced from the operating system. To my knowledge, `Ruby`, `curl`, and `go` do not even attempt to read the proxy bypass list from the operating system. And there is essentially no documentation related to the operating system's proxy bypass list on Windows and macOS. This is why I ran experiments with the operating system's default browser and attempted to test the "canonical" interpretation of the system proxy bypass list. From the results, it appears that macOS and Windows do not resolve the hostname.

**Comment by zooba** [comment]
The [`WINHTTP_PROXY_INFO` documentation](https://learn.microsoft.com/en-us/windows/win32/api/winhttp/ns-winhttp-winhttp_proxy_info) says that "proxy bypass list contains one or more server names".

Names are not what we get from a resolution process, so that would argue that at least on Windows, we should check the host name without resolving it.

Whether we can safely make that change in terms of backwards compatibility is a separate question. My instinct is that we should check both for now and plan to stop checking addresses after a deprecation period (and probably add some way for users to stop checking them now). But I'm not sure if that's best. @gpshead, I expect you know more about this area (proxies) than I do - any thoughts?

**Comment by weiiwang01** [comment]
@zooba Thank you for the inspiration on how to find the documentation. I discovered that the [corresponding documentation for macOS `kSCPropNetProxiesExceptionsList`](https://developer.apple.com/documentation/systemconfiguration/1517088-scdynamicstorecopyproxies) says that "Host name patterns that should bypass the proxy."

Does the "name" here imply that macOS shouldn't resolve the name as well?

**Comment by gpshead** [comment]
I agree with the majority of tools: A proxy bypass lists should always be a simple string match without name resolution. In proxy environments the point of the proxy is that it is entirely up to the proxy to do all resolving and remote connecting (or MITMing, denying, rejecting, etc). So we should not be doing _any_ name resolution (or reverse name lookups of an address) in order to determine if a proxy should be used.

@SethMichaelLarson may also have an opinion.

**Comment by ronaldoussoren** [comment]
> @zooba Thank you for the inspiration on how to find the documentation. I discovered that the [corresponding documentation for macOS `kSCPropNetProxiesExceptionsList`](https://developer.apple.com/documentation/systemconfiguration/1517088-scdynamicstorecopyproxies) says that "Host name patterns that should bypass the proxy."
> 
> Does the "name" here imply that macOS shouldn't resolve the name as well?

TLDR: Indeed, on macOS we shouldn't resolve the name to match system behaviour.

I did some testing with Apple's system libraries and those don't resolve names when determining if a proxy should be used.

That is, given the following script (requires PyObjC to be installed):

```python
import CFNetwork
import Cocoa
import SystemConfiguration
import socket


def main():
    hn = "www.bol.com"
    print(f"{hn} -> {socket.gethostbyname(hn)}")
    url = Cocoa.CFURLCreateWithString(None, f"http://{hn}/", None)
    print(f"{url=}")
    settings = SystemConfiguration.SCDynamicStoreCopyProxies(None)
    print(f"{settings=}")
    proxies = CFNetwork.CFNetworkCopyProxiesForURL(url, settings)
    print(f"{proxies=}")


if __name__ == "__main__":
    main()
```

I get the following output with a proxy configured an 34.35.121.47 added to the no-proxy list (the other two are part of the default configuration):

```
% python repro.py
www.bol.com -> 34.36.121.47
url=http://www.bol.com/
settings={
    ExceptionsList =     (
        "*.local",
        "169.254/16",
        "34.36.121.47"
    );
    FTPPassive = 1;
    HTTPEnable = 1;
    HTTPPort = 8080;
    HTTPProxy = "127.0.0.1";
    HTTPSEnable = 0;
    "__SCOPED__" =     {
        en0 =         {
            ExceptionsList =             (
                "*.local",
                "169.254/16",
                "34.36.121.47"
            );
            FTPPassive = 1;
            HTTPEnable = 1;
            HTTPPort = 8080;
            HTTPProxy = "127.0.0.1";
            HTTPSEnable = 0;
        };
    };
}
proxies=(
        {
        kCFProxyHostNameKey = "127.0.0.1";
        kCFProxyPortNumberKey = 8080;
        kCFProxyTypeKey = kCFProxyTypeHTTP;
    }
)
```

This shows that the "http://www.bol.com" will be proxied, even though its IP address is on the proxy list. 

That means that we shouldn't resolve names to match system behaviour.

Not that any of this is clearly documented (sigh...).

**Comment by weiiwang01** [comment]
@zooba I can update the code to print a deprecation warning when the hostname does not match directly, but the resolved result does, ensuring backward compatibility for now. However, based on the current evidence, I personally believe this should be categorized as a bug. How would you advise handling this issue?

**Comment by gpshead** [comment]
I agree with backporting this one, at least through bugfix releases.  I'd leave it to RMs to decide for older ones.

It hasn't been raised until now and client side behavior like this is not where real security gets implemented, but it is a possible information leak vs what most proxy configurations expects to happen.

I don't know how often proxy bypass lists actually get used in practice.  _(I'm used to the idea of all or nothing - but I haven't been within any application level proxy environments in decades)_

**Comment by weiiwang01** [comment]
> I agree with backporting this one, at least through bugfix releases. I'd leave it to RMs to decide for older ones.
> 
> It hasn't been raised until now and client side behavior like this is not where real security gets implemented, but it is a possible information leak vs what most proxy configurations expects to happen.
> 
> I don't know how often proxy bypass lists actually get used in practice. _(I'm used to the idea of all or nothing - but I haven't been within any application level proxy environments in decades)_

This problem affects almost all users who use the system proxy on macOS, as macOS automatically populates the proxy exception list with `*.local,169.254/16` by default, triggering this issue. On Windows, if a user selects the option to not use a proxy for local addresses, the proxy override list will include an `<local>` entry, which also triggers this problem.

BTW, is there anything I need to update in the pull request for backporting? I am using the `ipaddress` library, which is only available after 3.3. Is that okay, considering all releases before 3.3 are EOL? Sorry, this is my first time contributing here.

**Comment by gpshead** [comment]
> BTW, is there anything I need to update in the pull request for backporting?

Nope, don't worry about that, we'll do it when backporting as far as we decide (which wouldn't be further than 3.8 if we even go that far).

**Comment by gpshead** [comment]
[History] for posterity and to save others spelunking version control: The host resolution on proxy_bypass checking for Windows was originally added in 2001 by an unidentified contributor adding windows registry proxy bypass list lookup code: https://github.com/python/cpython/issues/33859 commit https://github.com/python/cpython/commit/55c12d4d5bd6adc8e60d074fbb5f1d666c87422c

There was no discussion about the name resolution at the time, it's just the code that got written.

The approach was probably copied when adding the equivalent macOS sysconf bypass code while de-Carbon-ing stuff to use modern APIs in ~2.6 in https://github.com/python/cpython/commit/9dd6b1db5be0293e9d359f7e75614c74b1e5e884.

**Comment by sethmlarson** [comment]
Thanks for the ping @gpshead, I'm +1 on removing the name resolution as it definitely represents an information leak if done on the originating machine instead of by the proxy. Also +1 for a backport into security releases.

**Comment by gpshead** [comment]
thanks!
